### PR TITLE
Improved error handling, simpler _transform function

### PR DESCRIPTION
**Pull request by mpcsh**
Using callbacks and the `iced-error` short-circuiter pattern, major refactor of `_transform`

### CODE REVIEW

File: src/chunk-stream.iced [hunk 1]
```diff
@@ -8,40 +9,62 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
-      @extra = null
+    if @writableObjectMode
+      await @transform_func(chunk, esc(defer(out)))
+      @push(out)
+      return cb(null)
 
-    # if we don't have a full block, dump everything into extra and skip this round
-    if chunk.length < @block_size
-      @extra = chunk
     else
```
Comment (oconnor663) [inline src/chunk-stream.iced]: We don't need the `else` anymore, since we're short-circuiting above.

File: src/chunk-stream.iced [hunk 2]
```diff
@@ -8,40 +9,62 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
-          chunk = chunk[...chunk.length-remainder]
-        @push(@transform_func(chunk))
+        # if we have to, call the transform function multiple times until we have less than a block size remaining
+        if @exact_chunking
+          while chunk.length >= @block_size
+            await @transform_func(chunk, esc(defer(out)))
+            @push(out)
+            chunk = chunk[@block_size...]
+          @extra = chunk
+          return cb(null)
```
Comment (oconnor663) [inline src/chunk-stream.iced]: Ditto. One of the goals of early exiting is to avoid the deep levels of indentation.

File: src/chunk-stream.iced [hunk 3]
```diff
@@ -8,40 +9,62 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
+
+      # either way, write out one final block (length guaranteed <= @block_size)
+      if @extra and @extra.length isnt 0
+        await @transform_func(@extra, esc(defer(out)))
+        @push(out)
+
+    await @_flush_append(esc(defer(out)))
+    if out? then @push(out)
+
+    return cb(null)
```
Comment (oconnor663) [inline src/chunk-stream.iced:53]: No need for `return` here at the very end.

File: src/chunk-stream.iced [hunk 4]
```diff
@@ -1,17 +1,26 @@
...
-  constructor : ({@transform_func, @block_size, @exact_chunking, @writableObjectMode, @readableObjectMode}) ->
+  constructor : ({@transform_func, @block_size, @readableObjectMode}) ->
     @extra = null
-    super({@writableObjectMode, @readableObjectMode})
+    super({@readableObjectMode})
+
+  # chunk must be guaranteed to be a multiple of the block size
+  _transform_chunk : (chunk, cb) ->
+    blocks = []
+    await
```
Comment (maxtaco) [inline src/chunk-stream.iced]: are you sure you want these done in parallel?

Comment (mpcsh) [reply src/chunk-stream.iced]: see ec02721 - it was `Buffer.concat` that was the bottleneck (we're now doing one big `concat` on the `blocks` array, instead of `concat`ing each block as we go along). I tested parallel vs serial with the same concat-an-array scheme and found no difference, so it's serial now for simplicity.

File: src/chunk-stream.iced [hunk 5]
```diff
@@ -1,17 +1,25 @@
...
+    super({@readableObjectMode})
+
+  # chunk must be guaranteed to be a multiple of the block size
+  _transform_chunk : (chunk, cb) ->
+    blocks = []
+    count = 0
+    for i in [0...chunk.length] by @block_size
+      block = chunk[i...i+@block_size]
+      await @transform_func(block, defer(err, blocks[count]))
+      ++count
```
Comment (oconnor663) [inline src/chunk-stream.iced]: Can we just use `array.push()` and get rid of `count`? (But maybe we should get rid of this entire function anyway? As per comment below.)

File: src/chunk-stream.iced [hunk 6]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
-        # chop off extra
-        if remainder isnt 0
-          @extra = chunk[chunk.length-remainder...]
-          chunk = chunk[...chunk.length-remainder]
-        @push(@transform_func(chunk))
-
-    cb()
+      return cb(null, new Buffer(''))
+
+    # next, mangle the chunk so that it is an even multiple of the block size
```
Comment (oconnor663) [inline src/chunk-stream.iced]: This section feels like an artifact of the exact chunking mode. Maybe just call `_transform_func` in a loop here, and get rid of the `_transform_chunk` function entirely?

File: src/chunk-stream.iced (hunk 5 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: Yes - before, when it was running in parallel, we couldn't - but now that it's in serial we can

File: src/chunk-stream.iced [hunk 7]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
-    unless @writableObjectMode
-      while @exact_chunking and @extra and @extra.length >= @block_size
-        @push(@transform_func(@extra[...@block_size]))
-        @extra = @extra[@block_size...]
-      if @extra and @extra.length isnt 0 then @push(@transform_func(@extra))
-    cb()
+    esc = make_esc(cb, "Failed to flush")
+
+    # potentially deal with multiple flushed blocks
+    if @extra?.length >= @block_size
```
Comment (oconnor663) [inline src/chunk-stream.iced]: Should this be impossible?

File: src/chunk-stream.iced [hunk 8]
```diff
@@ -1,17 +1,25 @@
...
-    super({@writableObjectMode, @readableObjectMode})
+    super({@readableObjectMode})
+
+  # chunk must be guaranteed to be a multiple of the block size
+  _transform_chunk : (chunk, cb) ->
+    blocks = []
+    count = 0
+    for i in [0...chunk.length] by @block_size
+      block = chunk[i...i+@block_size]
+      await @transform_func(block, defer(err, blocks[count]))
```
Comment (oconnor663) [inline src/chunk-stream.iced]: This `err` is dropped on the ground. (But maybe we should get rid of this entire function anyway? As per comment below.)

File: src/chunk-stream.iced (hunk 6 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: EDIT: After discussion, it is in fact a vestige of the old modes. Now calling in a loop and storing one potentially smaller block in `@extra`.

File: src/chunk-stream.iced [hunk 9]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
+      return cb(null, new Buffer(''))
+
+    # next, mangle the chunk so that it is an even multiple of the block size
+    remainder = chunk.length % @block_size
+    if remainder isnt 0
+      @extra = chunk[chunk.length-remainder...]
+      chunk = chunk[...chunk.length-remainder]
+
+    # finally, process the chunk (guaranteed to be the correct length)
+    esc = make_esc(cb, "Failed to transform the chunk")
```
Comment (oconnor663) [inline src/chunk-stream.iced]: By convention the `esc`'s string refers to the function name rather than the error condition itself. (Also it usually goes at the top of the function, unless you think there might be a good reason to put it lower?) If you want a useful message, I think you need to pass it into `new Error(...)` or similar.

File: src/chunk-stream.iced (hunk 7 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: Line 53? No. It'll probably never happen, but we can't make any guarantees.

File: src/chunk-stream.iced [hunk 10]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
+    await @_transform_chunk(chunk, esc(defer(out)))
+    cb(null, out)
+
+
+  # this is to be overridden by subclasses who want to output something else after the chunking is over
+  _flush_append : (cb) ->
```
Comment (oconnor663) [inline src/chunk-stream.iced]: Should we have a similar contract for `prepend`? Symmetry here might make sense.

File: src/chunk-stream.iced (hunk 8 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: Oh yeah, it totally is. Fixed. I like it factored out into its own function but it really doesn't make a difference

File: src/chunk-stream.iced (hunk 10 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: I don't think so - there's no analogue to it in the node interface. Would involve an `if` that would be unnecessary for most stream consumers.

File: src/chunk-stream.iced (hunk 9 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: Gotcha. Do you mean `esc = make_esc(cb, new Error('...'))`?

File: src/chunk-stream.iced [hunk 11]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
+    # potentially deal with multiple flushed blocks
+    if @extra?.length >= @block_size
+      await @_transform_chunk(@extra, esc(defer(out_blocks)))
+      if out_blocks?
+        @push(out_blocks)
+
+    # either way, write out one final block (length guaranteed <= @block_size)
+    if @extra?.length isnt 0
+      await @transform_func(@extra, esc(defer(out_short)))
+      if out_short?
```
Comment (oconnor663) [inline src/chunk-stream.iced]: Is `transform_func` allowed to return null in other places? Or only here?

File: src/chunk-stream.iced [hunk 12]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
+        @push(out_short)
+
+    # allow subclasses to push something out at the very very end
+    await @_flush_append(esc(defer(out_flush)))
+    if out_flush?
```
Comment (oconnor663) [inline src/chunk-stream.iced:50]: Ditto here. What does it mean to return null? If we're operating in object mode, and `null` is actually a valid output, is this a bug?

File: src/chunk-stream.iced [hunk 13]
```diff
@@ -20,28 +28,42 @@ exports.ChunkStream = class ChunkStream extends stream.Transform
...
-        @push(@transform_func(chunk))
-
-    cb()
+      return cb(null, new Buffer(''))
+
+    # next, mangle the chunk so that it is an even multiple of the block size
+    remainder = chunk.length % @block_size
+    if remainder isnt 0
+      @extra = chunk[chunk.length-remainder...]
+      chunk = chunk[...chunk.length-remainder]
```
Comment (oconnor663) [inline src/chunk-stream.iced]: From talk in person: this intermediate, partially-shrunk version of `chunk` probably doesn't need to exist.

File: src/chunk-stream.iced (hunk 9 above)
Comment (oconnor663) [reply src/chunk-stream.iced]: No, sorry. What I mean is that the `make_esc` call should probably go to the top of the function and look something like this:

```
esc = make_esc(cb, "ChunkStream::_transform")
```

Then if you decide that errors in `_transform` need to have extra context attached to them (that is, context that will make sense if it's printed in the terminal to a human), it could look something like this:

```
await @_transform_chunk(chunk, defer(err, out))
if err?
    return cb(new Error("Failed to transform the chunk: #{err}"), null)
```

That said, the chunk stream itself is probably an implementation detail that won't mean much to an end user. It might be enough to just rely on the `esc` stack trace here and return errors unmodified.

File: src/chunk-stream.iced (hunk 7 above)
Comment (oconnor663) [reply src/chunk-stream.iced]: From talk in person: we think this should be impossible. If we keep code in here for it, it should be an assert rather than real handling.

File: src/util.iced [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  esc = make_esc(cb, "Error in stream writing")
+  written = 0
+  expected_results = []
+  while written < len
+    # generate random length
+    await crypto.randomBytes(1, esc(defer(index)))
+    amt = (index[0] + 1)*16
+
+    # generate random bytes of length amt
+    await crypto.randomBytes(amt, esc(defer(buf)))
```
Comment (oconnor663) [inline src/util.iced:15]: I think we talked about this before, but random input in tests makes test failures harder to reproduce. It would be one thing if the randomness was really pulling its weight, in practice our random inputs don't seem to be hitting the corner cases we need. The number of bugs in https://github.com/keybase/node-saltpack/pull/2 that the tests didn't catch still has me worried.

File: src/chunk-stream.iced (hunk 7 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: After mulling this over and reading some node source code, it's definitely completely impossible. I'm fine with not handling it

File: src/chunk-stream.iced (hunk 12 above)
Comment (mpcsh) [reply src/chunk-stream.iced:50]: `@flush_append` returns null by default - you don't have to override it in subclasses. Even in object mode, `null` is not a valid output in node streams, it's reserved (which can be annoying).

File: src/chunk-stream.iced (hunk 13 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: Yep, done

File: src/chunk-stream.iced (hunk 11 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: Hm, no, it's definitely never allowed to call back with null

File: src/chunk-stream.iced [hunk 15]
```diff
@@ -1,17 +1,27 @@
...
 
 exports.ChunkStream = class ChunkStream extends stream.Transform
 
-  constructor : ({@transform_func, @block_size, @exact_chunking, @writableObjectMode, @readableObjectMode}) ->
+  constructor : ({@transform_func, @block_size, @readableObjectMode}) ->
     @extra = null
-    super({@writableObjectMode, @readableObjectMode})
+    super({@readableObjectMode})
+
+  _transform_chunk : (chunk, cb) ->
```
Comment (oconnor663) [inline src/chunk-stream.iced]: I think it's cleaner to get rid of this function and inline it. Normally I'm a big fan of breaking things up, but when the smaller functions have to remember intimate details about each other's implementations, it can hurt more than it helps. Note the implicit contract around `@extra` here. We need to know that it's empty on line 15, which depends on what's happening in the caller at lines 27-28.

Anyway, overall just a slight preference of mine. Your call.

**Comment by oconnor663** [comment]
LGTM, just the one thought above.

File: src/chunk-stream.iced (hunk 15 above)
Comment (mpcsh) [reply src/chunk-stream.iced]: I'm neutral at this point, so I'll go ahead and change it
